File: android/app/src/main/kotlin/io/rebble/cobble/handlers/AppInstallHandler.kt
package io.rebble.cobble.handlers

import android.content.Context
import io.rebble.cobble.datasources.WatchMetadataStore
import io.rebble.cobble.middleware.PutBytesController
import io.rebble.cobble.middleware.getBestVariant
import io.rebble.cobble.shared.handlers.CobbleHandler
import io.rebble.cobble.util.getAppPbwFile
import io.rebble.cobble.util.requirePbwAppInfo
import io.rebble.libpebblecommon.metadata.WatchHardwarePlatform
import io.rebble.libpebblecommon.packets.AppFetchRequest
import io.rebble.libpebblecommon.packets.AppFetchResponse
import io.rebble.libpebblecommon.packets.AppFetchResponseStatus
import io.rebble.libpebblecommon.services.AppFetchService
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.flow.first
import kotlinx.coroutines.launch
import kotlinx.coroutines.withTimeoutOrNull
import timber.log.Timber
import javax.inject.Inject

class AppInstallHandler @Inject constructor(
        coroutineScope: CoroutineScope,
        private val context: Context,
        private val appFetchService: AppFetchService,
        private val putBytesController: PutBytesController,
        private val watchMetadataStore: WatchMetadataStore
) : CobbleHandler {
    init {
        coroutineScope.launch {
            for (message in appFetchService.receivedMessages) {
                when (message) {
                    is AppFetchRequest -> onNewAppFetchRequest(message)
                }
            }
        }
    }

    private suspend fun onNewAppFetchRequest(message: AppFetchRequest) {
        try {
            val appUuid = message.uuid.get().toString()
            val appFile = getAppPbwFile(context, appUuid)

            if (!appFile.exists()) {
                respondFetchRequest(AppFetchResponseStatus.INVALID_UUID)
                Timber.e("Watch requested nonexistent app data (%s)", appUuid)
                return
            }

            if (putBytesController.status.value.state != PutBytesController.State.IDLE) {
                Timber.e("Watch requested new app data PutBytes is busy")
                respondFetchRequest(AppFetchResponseStatus.BUSY)
                return
            }

            // Wait some time for metadata to become available in case this has been called
            // Right after watch has been connected
            val hardwarePlatformNumber = withTimeoutOrNull(2_000) {
                watchMetadataStore.lastConnectedWatchMetadata.first { it != null }
            }
                    ?.running
                    ?.hardwarePlatform
                    ?.get()

            if (hardwarePlatformNumber == null) {
                Timber.e("No watch metadata available. Cannot deduce watch type.")
                respondFetchRequest(AppFetchResponseStatus.NO_DATA)
                return
            }


            val connectedWatchType = WatchHardwarePlatform
                    .fromProtocolNumber(hardwarePlatformNumber)
                    ?.watchType

            if (connectedWatchType == null) {
                Timber.e("Unknown hardware platform %s", hardwarePlatformNumber)
                respondFetchRequest(AppFetchResponseStatus.NO_DATA)
                return
            }

            val appInfo = requirePbwAppInfo(appFile)

            val targetWatchType = getBestVariant(connectedWatchType, appInfo.targetPlatforms)
            if (targetWatchType == null) {
                Timber.e("Watch %s is not compatible with app %s Compatible apps: %s",
                        targetWatchType,
                        appUuid,
                        appInfo.targetPlatforms
                )
                respondFetchRequest(AppFetchResponseStatus.NO_DATA)
                return
            }


            respondFetchRequest(AppFetchResponseStatus.START)
            putBytesController.startAppInstall(message.appId.get(), appFile, targetWatchType)
        } catch (e: Exception) {
            Timber.e(e, "AppFetch fail")
            respondFetchRequest(AppFetchResponseStatus.NO_DATA)

        }
    }

    private suspend fun respondFetchRequest(status: AppFetchResponseStatus) {
        appFetchService.send(AppFetchResponse(status))
    }
}

File: ios/Runner/PebbleProtocol/AppInstallHandler.swift
//
//  AppInstallHandler.swift
//  Runner
//
//  Created by crc32 on 26/02/2022.
//

import Foundation
import libpebblecommon
import CocoaLumberjackSwift
import PromiseKit

class AppInstallHandler {
    private let appFetchService: AppFetchService
    
    init(appFetchService: AppFetchService) {
        self.appFetchService = appFetchService
        appFetchService.receivedMessages.receive(completionHandler: onAppFetchServiceMessage)
    }
    
    func onAppFetchServiceMessage(message: Any?, e: Error?) {
        defer {
            appFetchService.receivedMessages.receive(completionHandler: onAppFetchServiceMessage)
        }
        if (message is AppFetchRequest) {
            let message: AppFetchRequest = message as! AppFetchRequest
            let appUuid = message.uuid.get()
            do {
                let appFile = try getAppPbwFile(appUuid: appUuid?.description() ?? "")
                guard FileManager.default.fileExists(atPath: appFile.path) else {
                    respondFetchRequest(status: .invalidUuid).cauterize()
                    DDLogError("Watch requested nonexistent app data \(appUuid?.description() ?? "nil")")
                    return
                }
                
                guard ProtocolComms.shared.putBytesController.status.state == .Idle else {
                    DDLogError("Watch requested new app data but PutBytes is busy")
                    respondFetchRequest(status: .busy).cauterize()
                    return
                }
                
                guard let hardwarePlatformNumber = WatchMetadataStore.shared.lastConnectedWatchMetadata?.running.hardwarePlatform.get()?.uint8Value else {
                    DDLogError("No watch metadata available. Cannot deduce watch type.")
                    respondFetchRequest(status: .noData).cauterize()
                    return
                }
                
                guard let connectedWatchType = WatchHardwarePlatform.companion.fromProtocolNumber(number: hardwarePlatformNumber) else {
                    DDLogError("Unknown hardware platform \(hardwarePlatformNumber)")
                    respondFetchRequest(status: .noData).cauterize()
                    return
                }
                
                let appInfo = try requirePbwAppInfo(pbwFile: appFile)
                guard let targetWatchType = connectedWatchType.watchType.getBestVariant(availableAppVariants: appInfo.targetPlatforms) else {
                    DDLogError("Watch \(connectedWatchType) is not compatible with app \(String(describing: appUuid)) Compatible platforms: \(appInfo.targetPlatforms)")
                    respondFetchRequest(status: .noData).cauterize()
                    return
                }
                
                self.respondFetchRequest(status: .start).then { () -> Promise<Void> in
                    guard let appId = message.appId.get()?.uintValue else {
                        DDLogError("appID was null")
                        return self.respondFetchRequest(status: .noData)
                    }
                    return try ProtocolComms.shared.putBytesController.startAppInstall(appId: appId, pbwFile: appFile, watchType: targetWatchType).asVoid()
                }.catch { error in
                    DDLogError("Exception while catering to to AppFetchService message:" + ((error as? CobbleError)?.message ?? error.localizedDescription))
                    if (error is PbwSpecError) {
                        self.respondFetchRequest(status: .noData).cauterize()
                    }
                }
            } catch {
                DDLogError("Exception while handling AppFetchService message:" + error.localizedDescription)
            }
        }
    }
    
    private func respondFetchRequest(status: AppFetchResponseStatus) -> Promise<Void> {
        return Promise<Void> { (seal) throws -> () in
            appFetchService.send(packet: AppFetchResponse(status: status)) {_,e in
                guard e == nil else {
                    DDLogError("Failed to respond to AppFetchRequest: " + String(describing: e!))
                    seal.reject(e!)
                    return
                }
                DDLogDebug("Responded with: \(status.name)")
                seal.fulfill(())
            }
        }
        
    }
}
